### DISCUSSION TITLE
Composite Primary key

### DISCUSSION THREAD

**Discussion by DataTowerSystems** [Finbuckle/Finbuckle.MultiTenant#555] [Q&A]
First of all thank you very much for this greate library.
My case is as following:
I have 2 tables (Entities) Product and Category
The product in domain:

public class Product : AuditableEntity, IAggregateRoot
{
    public string ProductCode { get; set; } = string.Empty;
    public string ProductDescription { get; set; } = string.Empty;
    public string ProductCategoryCode { get; set; } = string.Empty;
    public Category Category { get; set; } = default!;

    [NotMapped]
    public string? TenantId { get; set; }

    public Product ClearImagePath()
    {
        ImagePath = string.Empty;
        return this;
    }
}

And the Category in Domain:

public class Category : AuditableEntity, IAggregateRoot
{
    public string CategoryCode { get; set; } = string.Empty;
    public string CategoryDescriptionArabic { get; set; } = string.Empty;
    public string CategoryDescriptionEnglish { get; set; } = string.Empty;
    [NotMapped]
    public string? TenantId { get; set; }
    public ICollection<Product> Products { get; set; } = default!;
}

And in Infrastructure/ Persistence/ Configuration
The fluent API

public class CategoryConfig : IEntityTypeConfiguration<Category>
{
    public void Configure(EntityTypeBuilder<Category> builder)
    {
        builder.IsMultiTenant();

        builder
            .HasIndex(b => new { b.CategoryCode, b.TenantId })
            .IsUnique(true)
            .HasDatabaseName("IX_CategoryCode_TenantId");

        builder
            .Property(b => b.CategoryCode)
            .IsUnicode(false)
            .HasMaxLength(50)
     }
}

public class ProductConfig : IEntityTypeConfiguration<Product>
{
    public void Configure(EntityTypeBuilder<Product> builder)
    {
        builder.IsMultiTenant();

        builder
        .HasIndex(b => new { b.ProductCode, b.TenantId })
        .IsUnique(true);

       <---- This is the relation between the Product and the Category ------->
          **builder
            .HasOne(b => b.Category)
            .WithMany(b => b.Products)
            .HasForeignKey(b => new { b.ProductCategoryCode, b.TenantId })
            .HasPrincipalKey(b => new { b.CategoryCode, b.TenantId })
            .IsRequired();**
       <---- This is the relation between the Product and the Category ------->
    }
}

And this is how it looks in the Database

relations

Until now everything is Ok, the problem is when I create new Category and try to save it I got the following error:
“Unable to track an entity of type 'Category' because alternate key property 'TenantId' is null. If the alternate key is not used in a relationship, then consider using a unique index instead. Unique indexes may contain nulls, while alternate keys may not.”

I hope this is clear now.
Can you please help to solve this issue.
Again Thank you very much for your cooperation.

**Comment by AndrewTriesToCode** [comment]
Hi! I'm glad you find Finbuckle helpful! I'm at a conference today so I can't give you a full answer until later today, but I will make sure to respond.

**Reply by DataTowerSystems** [reply]
> Hi! I'm glad you find Finbuckle helpful! I'm at a conference today so I can't give you a full answer until later today, but I will make sure to respond.

Thank you very much.

**Reply by DataTowerSystems** [reply]
Please I am waiting for your answer.
Thank you.

**Comment by AndrewTriesToCode** [comment]
Do you have a simple repo that shows this problem you can link me to?

Is the error happening when `SaveChanges` is called or when adding the entity to the dbcontext? Can you shows the code where you are adding the category?

**Reply by DataTowerSystems** [reply]
I am using this https://github.com/fullstackhero/dotnet-webapi-boilerplate which it use Finbuckle library for Multi tenancy.
This is the saving task in the BaseDbContext class
```
 public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken = new CancellationToken())
    {
        var auditEntries = HandleAuditingBeforeSaveChanges(_currentUser.GetUserId());

        int result = await base.SaveChangesAsync(cancellationToken);

        await HandleAuditingAfterSaveChangesAsync(auditEntries, cancellationToken);

        await SendDomainEventsAsync();

        return result;
    }
```
The error is when trying to save the row with the Alternate Key says:
"Alternate key can not be null" because the TenantId in that row is null 
```
<---- This is the relation between the Product and the Category ------->
      **builder
        .HasOne(b => b.Category)
        .WithMany(b => b.Products)
        .HasForeignKey(b => new { b.ProductCategoryCode, b.TenantId })
        .HasPrincipalKey(b => new { b.CategoryCode, b.TenantId })
        .IsRequired();**
   <---- This is the relation between the Product and the Category ------->
```
I hope that I make it clear, thank you very much.

**Reply by ashishsinha24** [reply]
@DataTowerSystems 

Even I am facing the same problem. Did you find a solution for composite keys?

Thanks
